### PR TITLE
Group subcommands with debug markers

### PR DESCRIPTION
**Pull request by Qining**
Use CmdIDGroup in SubCmdRoot's SubGroup to group the subcommands

### CODE REVIEW

File: gapis/api/cmd_id_group.go [hunk 1]
```diff
@@ -60,6 +61,14 @@ func NewRoot(idx []uint64) *SubCmdRoot {
 		SubGroup: CmdIDGroup{Name: "Subgroup"}}
 }
 
+// SubCmdMarkerGroup contains the range info of a marker group of subcommands.
+type SubCmdMarkerGroup struct {
```
Comment (Qining) [inline gapis/api/cmd_id_group.go]: Or maybe we can just use `CmdIDGroup` directly?

Comment (ben-clayton) [reply gapis/api/cmd_id_group.go]: F2F discussion: Will be removed.

File: gapis/api/cmd_id_group.go [hunk 2]
```diff
@@ -221,6 +230,12 @@ func (g CmdIDGroup) Index(index uint64) SpanItem {
 	for _, s := range g.Spans {
 		c := s.itemCount()
 		if index < c {
+			// If the CmdIDGroup is a subcommand group, need to combine the SubCmdIdx
+			// of the immediate parent SubCmdRoot of this CmdIDGroup with the index
+			// target.
+			if x, ok := s.item(index).(SubCmdIdx); ok {
+				return append(g.Parent, x...)
```
Comment (ben-clayton) [inline gapis/api/cmd_id_group.go]: F2F discussion: This is mixing relative with absolute. Please just be relative (no change).

File: gapis/api/cmd_id_group.go [hunk 3]
```diff
@@ -307,24 +322,27 @@ func (g CmdIDGroup) IterateBackwards(index uint64, cb func(childIdx uint64, item
 // All groups must be added before commands.
 // Attemping to call this function after commands have been added may result in
 // panics!
-func (g *CmdIDGroup) AddGroup(start, end CmdID, name string) error {
+func (g *CmdIDGroup) AddGroup(start, end CmdID, name string) (*CmdIDGroup, error) {
```
Comment (ben-clayton) [inline gapis/api/cmd_id_group.go]: F2F discussion: These changes are also done by one of my CLs. I will put this change up for review ASAP.

File: gapis/api/cmd_id_group.go [hunk 4]
```diff
@@ -413,18 +432,63 @@ func (c *SubCmdRoot) Insert(base []uint64, r []uint64) {
 		}
 		sg.Insert(append(base, r[0]), r[1:])
 	} else {
-		if len(c.SubGroup.Spans) == 0 {
-			c.SubGroup.Spans = Spans{&CmdIDRange{CmdID(0), id + 1}}
+		// Add subcommand one by one to the sub groups of the SubCmdRoot
```
Comment (ben-clayton) [inline gapis/api/cmd_id_group.go]: OK

Comment (ben-clayton) [reply gapis/api/cmd_id_group.go]: please can we have better docs on this function.

File: gapis/api/cmd_id_group.go [hunk 5]
```diff
@@ -413,18 +432,63 @@ func (c *SubCmdRoot) Insert(base []uint64, r []uint64) {
...
-			r.End = id + 1
-		} else {
-			panic("This should not happen, inner-most subcommands must be added to SubCmdRoot before other nesting Sub-SubCmdRoot")
+		for i := CmdID(0); i <= id; i++ {
+			c.SubGroup.AddCommand(i)
 		}
 	}
 }
 
+func (c *SubCmdRoot) AddSubCmdMarkerGroups(base []uint64, groups []*SubCmdMarkerGroup) error {
```
Comment (ben-clayton) [inline gapis/api/cmd_id_group.go]: F2F discussion: 
* Documentation
* `base` parameter should be relative to c, removes the `c.Id.Contains(SubCmdIdx(base))` condition.

File: gapis/api/cmd_id_group.go [hunk 6]
```diff
@@ -333,23 +351,24 @@ func (g *CmdIDGroup) AddGroup(start, end CmdID, name string) error {
...
-			n := CmdIDGroup{Name: name, Range: r, Spans: make(Spans, c)}
-			copy(n.Spans, g.Spans[s:s+c])
-			slice.Replace(&g.Spans, s, c, &n)
+			newGroup = &CmdIDGroup{Name: name, Range: r, Spans: make(Spans, c)}
+			copy(newGroup.Spans, g.Spans[s:s+c])
+			slice.Replace(&g.Spans, s, c, newGroup)
 		}
 	}
-	return nil
+	return newGroup, err
```
Comment (ben-clayton) [inline gapis/api/cmd_id_group.go]: F2F discussion:

Just a common thing that's tripped us up in the past:

```go
if err != nil {
    return nil, err
}
return newGroup, nil
```

**Comment by Qining** [comment]
Migrate to #1064 . All comments here will be resolved in #1061

File: gapis/api/cmd_id_group.go (hunk 5 above)
Comment (Qining) [reply gapis/api/cmd_id_group.go]: Fixed in 6b24e7f

File: gapis/api/cmd_id_group.go (hunk 1 above)
Comment (Qining) [reply gapis/api/cmd_id_group.go]: Fixed in Fixed in #1064

File: gapis/api/cmd_id_group.go (hunk 2 above)
Comment (Qining) [reply gapis/api/cmd_id_group.go]: Fixed in #1064

File: gapis/api/cmd_id_group.go (hunk 3 above)
Comment (Qining) [reply gapis/api/cmd_id_group.go]: Not yet updated, will wait for rebase

File: gapis/api/cmd_id_group.go (hunk 4 above)
Comment (Qining) [reply gapis/api/cmd_id_group.go]: Fixed in #1064

**Pull request closed without merging**